diff --git twisted/conch/ssh/session.py twisted/conch/ssh/session.py
index c50372c..e315c8f 100755
--- twisted/conch/ssh/session.py
+++ twisted/conch/ssh/session.py
@@ -188,6 +188,7 @@ class SSHSessionProcessProtocol(protocol.ProcessProtocol):
 
     def __init__(self, session):
         self.session = session
+        self.lostOutOrErrFlag = False
 
     def connectionMade(self):
         if self.session.buf:
@@ -200,8 +201,20 @@ class SSHSessionProcessProtocol(protocol.ProcessProtocol):
     def errReceived(self, err):
         self.session.writeExtended(connection.EXTENDED_DATA_STDERR, err)
 
-    def inConnectionLost(self):
-        self.session.conn.sendEOF(self.session)
+    def outConnectionLost(self):
+        """
+        EOF should only be sent when both STDOUT and STDERR have been closed.
+        """
+        if self.lostOutOrErrFlag is True:
+            self.session.conn.sendEOF(self.session)
+        else:
+            self.lostOutOrErrFlag = True
+
+    def errConnectionLost(self):
+        """
+        See outConnectionLost().
+        """
+        self.outConnectionLost()
 
     def connectionLost(self, reason = None):
         self.session.loseConnection()
diff --git twisted/conch/test/test_session.py twisted/conch/test/test_session.py
index f59fd72..e901913 100644
--- twisted/conch/test/test_session.py
+++ twisted/conch/test/test_session.py
@@ -1113,11 +1113,25 @@ class SSHSessionProcessProtocolTestCase(unittest.TestCase):
                 [(1, 'test data')])
 
 
-    def test_inConnectionLost(self):
+    def test_outConnectionLost(self):
         """
-        When inConnectionLost is called, it should send an EOF message,
+        When outConnectionLost and errConnectionLost are both called, we should
+        send an EOF message.
         """
-        self.pp.inConnectionLost()
+        self.pp.outConnectionLost()
+        self.assertFalse(self.session in self.session.conn.eofs)
+        self.pp.errConnectionLost()
+        self.assertTrue(self.session.conn.eofs[self.session])
+
+
+    def test_errConnectionLost(self):
+        """
+        Make sure reverse ordering of events in test_outConnectionLost also
+        sends EOF.
+        """
+        self.pp.errConnectionLost()
+        self.assertFalse(self.session in self.session.conn.eofs)
+        self.pp.outConnectionLost()
         self.assertTrue(self.session.conn.eofs[self.session])
 
 
